### PR TITLE
Improves the response banks

### PR DESCRIPTION
**Pull request by flaviens**
This commit contains three features that substantially improve the response banks:
* It fixes issues occurring when the burst length is not systematically the maximal possible burst length. The design was too simplistic and did not manage these cases properly.
* Instead of storing successive burst data in width, it now stores them in depth in the RAM.
* It provides support for dynamically dropping release enable signals (before, required the release enable signal to drop one cycle in advance).

### CODE REVIEW

File: rtl/simmem_resp_bank.sv [hunk 1]
```diff
@@ -65,15 +65,57 @@
...
+//  *  | Burst id 2, data 2.      |  Addr: 2, full addr: 10.
+//  *  +--------------------------+
+//  *  | Burst id 2, data 3.      |  Addr: 2, full addr: 11.
+//  *  +--------------------------+
+//
 
 module simmem_resp_bank #(
-    parameter int unsigned MaxBurstLen = 4,
     parameter int unsigned TotCapa = simmem_pkg::ReadDataBankCapacity,
+    parameter int unsigned MaxBurstLen = simmem_pkg::MaxRBurstLen,
```
Comment (vogelpi) [inline rtl/simmem_resp_bank.sv]: `MaxRBurstLen` does not exist.

File: rtl/simmem_resp_bank.sv [hunk 2]
```diff
@@ -65,15 +65,57 @@
...
-    localparam int unsigned DataWidth = $bits (DataType)  // derived parameter
+    // BurstLenWidth is the bit width to encode up to MaxBurstLen-1 in binary format.
+    //
+    // XBurstLenWidth is the bit width to encode up to MaxBurstLen in binary format. This is useful,
+    // as any RAM cell can contain or reserve 0, 1, ... or MaxBurstLen burst data.
+    localparam int unsigned XBurstLenWidth = $clog2 (MaxBurstLen + 1),  // derived parameter
+    localparam int unsigned BurstLenWidth = MaxBurstLen == 1 ? 1 : $clog2 (MaxBurstLen),  // derived parameter
+    localparam int unsigned BankAddrWidth = $clog2 (TotCapa),   // derived parameter
+    localparam int unsigned DataWidth = $bits (DataType),       // derived parameter
+    localparam int PayloadRamDepth = TotCapa * MaxBurstLen      // derived parameter
```
Comment (vogelpi) [inline rtl/simmem_resp_bank.sv]: `int unsigned`?

File: rtl/simmem_resp_bank.sv [hunk 3]
```diff
@@ -65,15 +65,57 @@
...
 
-    localparam int unsigned BurstLenWidth = $clog2 (MaxBurstLen + 1),  // derived parameter
-    localparam int unsigned BankAddrWidth = $clog2 (TotCapa),  // derived parameter
-    localparam int unsigned DataWidth = $bits (DataType)  // derived parameter
+    // BurstLenWidth is the bit width to encode up to MaxBurstLen-1 in binary format.
+    //
+    // XBurstLenWidth is the bit width to encode up to MaxBurstLen in binary format. This is useful,
+    // as any RAM cell can contain or reserve 0, 1, ... or MaxBurstLen burst data.
+    localparam int unsigned XBurstLenWidth = $clog2 (MaxBurstLen + 1),  // derived parameter
+    localparam int unsigned BurstLenWidth = MaxBurstLen == 1 ? 1 : $clog2 (MaxBurstLen),  // derived parameter
```
Comment (vogelpi) [inline rtl/simmem_resp_bank.sv]: Be careful: In AXI, the burst length is number of data beats - 1. So one data beat means burst len = 0, two data beats -> burst len = 1.

Comment (vogelpi) [reply rtl/simmem_resp_bank.sv]: In that sense, I would try to avoid the distinction between `BurstLenWidth` and `XBurstLenWidth`.

File: rtl/simmem_resp_bank.sv [hunk 4]
```diff
@@ -105,13 +147,17 @@ module simmem_resp_bank #(
 
   // Interface with the requester
   input  logic out_rsp_ready_i,
-  output logic out_rsp_valid_o
+  output logic out_rsp_valid_o,
+
+  // Ready signal from the delay calculator
+  input  logic delay_calc_ready_i,
+  // Ready signal to the delay calculator
+  output logic delay_calc_ready_o
```
Comment (vogelpi) [inline rtl/simmem_resp_bank.sv]: This change does not seem to be related to the other work in that PR.

File: rtl/simmem_resp_bank.sv [hunk 5]
```diff
@@ -65,15 +65,57 @@
 //    pointer a must be updated with the same value. Piggyback is a lazy operation (in the sense
 //    that if the pointer b is not updated, then a remains untouched as well) that maintains the
 //    pointer ordering.
+//
+//  Burst data storage: To make burst data storage transparent, we introduce the notion of extended
+//    RAM cell (as opposed to plain "elementary" RAM cells), only applicable and useful for the
+//    payload RAM. An extended payload RAM cell is a succession of MaxBurstLen elementary RAM cells,
+//    aligned to a MaxBurstLen "full" address. We define the full address as the address pointing to
+//    a given elementary cell, while a ("plain") address refers to the index of the extended cell.
+//    This makes addresses in metadata RAMs match with "plain" addresses of the payload RAM.
```
Comment (vogelpi) [inline rtl/simmem_resp_bank.sv]: Thanks for the useful comment :+1:

File: rtl/simmem_resp_bank.sv [hunk 6]
```diff
@@ -203,14 +250,17 @@ module simmem_resp_bank #(
   logic is_rsp_head_emptybox_q[NumIds];
 
   // Lengths of reservation and response queues.
-  logic [BankAddrWidth-1:0] rsv_len_d[NumIds];
-  logic [BankAddrWidth-1:0] rsv_len_q[NumIds];
+  logic [BankAddrWidth:0] rsv_len_d[NumIds];
+  logic [BankAddrWidth:0] rsv_len_q[NumIds];
```
Comment (vogelpi) [inline rtl/simmem_resp_bank.sv]: Please don't do this! If you really need this additional bit (do you?) please create a new localparam `LenWidth` that is `BankAddrWidth+1` and place a comment why you need one more bit in the length signals.

File: rtl/simmem_resp_bank.sv [hunk 7]
```diff
@@ -298,81 +372,145 @@ module simmem_resp_bank #(
...
+    assign
+    // The response length is decreased after output if there is an output for this identifier
+    rsp_len_after_out[i_id] = out_rsp_valid_o && out_rsp_ready_i && cur_out_id_onehot[i_id] &&
+    // And the burst data will be completely empty in the extended payload RAM cell under the tail
+    // pointer.
+    tail_rsv_cnt_id[i_id] == 0 && tail_rsp_cnt_id[i_id] == 0 && !cnt_in_mask_id[i_id] ?
+        rsp_len_q[i_id] - 1 : rsp_len_q[i_id];
+
+    // A linked list is said to await data when both its tail and pre_tail response bursts are
+    // empty. In the opposite case, the linked list could provide data.
```
Comment (vogelpi) [inline rtl/simmem_resp_bank.sv:519]: What do you mean by "tail and pre_tail response bursts are empty?" Please check the language here.

File: rtl/simmem_resp_bank.sv [hunk 8]
```diff
@@ -298,81 +372,145 @@ module simmem_resp_bank #(
...
 
+  // Calculate the length of each AXI identifier queue after the potential output
+  for (genvar i_id = 0; i_id < NumIds; i_id = i_id + 1) begin : gen_len_after_output
+    assign
+    // The response length is decreased after output if there is an output for this identifier
+    rsp_len_after_out[i_id] = out_rsp_valid_o && out_rsp_ready_i && cur_out_id_onehot[i_id] &&
+    // And the burst data will be completely empty in the extended payload RAM cell under the tail
+    // pointer.
+    tail_rsv_cnt_id[i_id] == 0 && tail_rsp_cnt_id[i_id] == 0 && !cnt_in_mask_id[i_id] ?
+        rsp_len_q[i_id] - 1 : rsp_len_q[i_id];
```
Comment (vogelpi) [inline rtl/simmem_resp_bank.sv]: This needs to be rewritten with a more clear formatting and comment. For example:
```
// Decrease response length if we are outputting for the the corresponding identifier and
rsp_len_after_out[i_id] = (out_rsp_valid_o && out_rsp_ready_i && cur_out_id_onehot[i_id] &&
                           // if the current output is the last data beat of the burst.
                           tail_rsv_cnt_id[i_id] == 0 &&
                           tail_rsp_cnt_id[i_id] == 0 &&
                           !cnt_in_mask_id[i_id]) ? rsp_len_q[i_id] - 1 : rsp_len_q[i_id];
```

Comment (vogelpi) [reply rtl/simmem_resp_bank.sv]: Related to that, isn't it possible to know when we are receiving/transmitting the last data beat without all these counters?

File: rtl/simmem_resp_bank.sv [hunk 9]
```diff
@@ -596,27 +696,45 @@ module simmem_resp_bank #(
...
-  end : gen_out_data
 
+  ///////////////////////////////////////////
+  // Payload RAM address offset management //
+  ///////////////////////////////////////////
+
+  // Until this step, payload data is managed as if there had been one payload RAM address per full
+  // response burst. To store multiple data of the same burst, however, the module uses consecutive
+  // elementary RAM cells. In this part, the payload RAM access addresses are generated as follows:
+  // realAddress = (iid << MaxBurstLen) + offset.
```
Comment (vogelpi) [inline rtl/simmem_resp_bank.sv]: Very clear, thanks!

File: rtl/simmem_resp_bank.sv [hunk 10]
```diff
@@ -813,41 +937,42 @@ module simmem_resp_bank #(
...
+                               pre_tail_rsp_cnt_id[i_id]);
           end else begin
             pyld_ram_out_addr_id[i_id] = tails[i_id];
-            // Set the corresponding payload RAM output wmask bit to one.
-            for (int unsigned i_bur = 0; i_bur < MaxBurstLen; i_bur = i_bur + 1) begin
-              pyld_ram_out_wmask_id[i_id][i_bur] = tail_cnt_id[i_id] ==
-                  MaxBurstLen[BurstLenWidth - 1:0] - i_bur[BurstLenWidth - 1:0] + 1;
-            end
+            pyld_ram_out_offset_id[i_id] = BurstLenWidth
+                '(tail_burst_len_id[i_id] - tail_rsv_cnt_id[i_id] - tail_rsp_cnt_id[i_id]);
```
Comment (vogelpi) [inline rtl/simmem_resp_bank.sv]: Is this correct syntax? Please make sure the `'` is on the same line as `BurstLenWidth`.

File: rtl/simmem_resp_bank.sv [hunk 11]
```diff
@@ -813,41 +937,42 @@ module simmem_resp_bank #(
...
           end
         end else begin
           pyld_ram_out_addr_id[i_id] = tails[i_id];
-          // Set the corresponding payload RAM output wmask bit to one.
-          for (int unsigned i_bur = 0; i_bur < MaxBurstLen; i_bur = i_bur + 1) begin
-            pyld_ram_out_wmask_id[i_id][i_bur] =
-                tail_cnt_id[i_id] == MaxBurstLen[BurstLenWidth - 1:0] - i_bur[BurstLenWidth - 1:0];
-          end
+          pyld_ram_out_offset_id[i_id] = BurstLenWidth
+              '(tail_burst_len_id[i_id] - tail_rsv_cnt_id[i_id] - tail_rsp_cnt_id[i_id]);
```
Comment (vogelpi) [inline rtl/simmem_resp_bank.sv]: Same comment here.

File: rtl/simmem_resp_bank.sv [hunk 12]
```diff
@@ -813,41 +937,42 @@ module simmem_resp_bank #(
...
+        // If this is the first data to accept in the burst, then update the response length, as the
+        // response length translates how many extended cells in the linked list contain at least
+        // one data to release, or has contained data of a burst whose data has not all been
+        // released (and possibly not received) yet.
+        if (rsp_burst_len_id[i_id] == rsp_rsv_cnt_id[i_id]) begin
           rsp_len_d[i_id] = rsp_len_d[i_id] + 1;
+        end
+        // If this is the last data to accept in the burst, then update the head pointers.
+        // rsp_burst_len_id[i_id] - rsp_rsv_cnt_id[i_id] corresponds to the remaining burst data to
+        // receive. comment
```
Comment (vogelpi) [inline rtl/simmem_resp_bank.sv]: comment?

File: rtl/simmem_resp_bank.sv (hunk 8 above)
Comment (vogelpi) [reply rtl/simmem_resp_bank.sv]: I thought more about this. The problem I see with these burst counters `rsv_cnt_q`, `burst_len_q` and `rsp_cnt_q` is that there are `TotCapa` many of them and each one `XBurstLenWidth` bits wide. The number of FFs will grow very quickly for larger configurations.

But I think you cannot do without them or it's not worth the design effort. What you would really need is one counter per AXI ID/linked list that you set to the burst length (could be stored in a metadata ram) when receiving the first data beat of the response. Then you would count down and set and when you reach the second to last beat, you write the `last` bit in the payload RAM. Then when outputting the data from the payload RAM you wouldn't need to do any counting anymore. Just watch for the `last` (or second to last bit). When you see this, you know the next beat is the last one - time to update your pointers...

But as I said, it's probably not worth the effort. Especially since you already use the same counters also in different places. 


and think it would possible to do without burst counters:

**Review by vogelpi — COMMENTED** [review]
Hi @flaviens ,
this looks quite okay. I have left a few comments that need fixing before we can merge it.

Best,
Pirmin

File: rtl/simmem_resp_bank.sv (hunk 4 above)
Comment (flaviens) [reply rtl/simmem_resp_bank.sv]: This is true. Should I remove it and add it later?

File: rtl/simmem_resp_bank.sv (hunk 8 above)
Comment (flaviens) [reply rtl/simmem_resp_bank.sv]: True, I had not thought enough before writing that apparently. There seems, indeed, to be much more efficient ways to implement the same thing.  Should I do this in this PR or prepare for a new PR? It could be seen as an improvement on this one.

File: rtl/simmem_resp_bank.sv (hunk 10 above)
Comment (flaviens) [reply rtl/simmem_resp_bank.sv]: This has been automatically formatted, and accepted by Verilator and Vivado, but my syntax coloring has some difficulties to recognize it. I'll change this.

File: rtl/simmem_resp_bank.sv (hunk 12 above)
Comment (flaviens) [reply rtl/simmem_resp_bank.sv]: Oops :) thanks! The comment was out of date and I had noted that I had to update it. The _TODO_ disappeared somehow, deleted it by mistake without solving it.

File: rtl/simmem_resp_bank.sv (hunk 3 above)
Comment (flaviens) [reply rtl/simmem_resp_bank.sv]: Thanks! I had not noticed this. I'll see how I can correct this.

File: rtl/simmem_resp_bank.sv (hunk 1 above)
Comment (flaviens) [reply rtl/simmem_resp_bank.sv]: Thanks, I'll add the simmem_pkg into this PR. I also changed to a unique MaxBurstLen for read and write, as a separate does not bring that much and makes really cumbersome expressions (see comments below).

File: rtl/simmem_resp_bank.sv (hunk 2 above)
Comment (flaviens) [reply rtl/simmem_resp_bank.sv]: Yes, changed, thanks!

File: rtl/simmem_resp_bank.sv (hunk 3 above)
Comment (flaviens) [reply rtl/simmem_resp_bank.sv]: Actually your comment here was of really high value, because that was a large mistake that I had made, and a smaller one:
* The smaller: I re-read my own code too fast when I decided to add the +1 to the counters (see further below), because actually this was already managed by the XBurstLenWidth. This is why this XBurstLenWidth appeared useless for my last push.
* The larger: I had not made the difference between the burst length field value and the actual burst length (and same for the burst size).

Both are corrected now debugged, and tested. I tried to make the code as digest as possible but it adds a little bit of complexity, as I wanted to keep the (smaller) burst length field value wherever possible, and "decompress" it (bit shift) only when necessary.

File: rtl/simmem_resp_bank.sv (hunk 4 above)
Comment (vogelpi) [reply rtl/simmem_resp_bank.sv]: Yes, please.

File: rtl/simmem_resp_bank.sv (hunk 8 above)
Comment (vogelpi) [reply rtl/simmem_resp_bank.sv]: I would do a follow up PR.

File: rtl/simmem_resp_bank.sv (hunk 10 above)
Comment (vogelpi) [reply rtl/simmem_resp_bank.sv]: Thanks.

File: rtl/simmem_resp_bank.sv (hunk 1 above)
Comment (vogelpi) [reply rtl/simmem_resp_bank.sv]: Okay, sounds good.

File: rtl/simmem_pkg.sv [hunk 13]
```diff
@@ -4,78 +4,169 @@
...
+  parameter int unsigned WStrbWidth = MaxBurstEffSizeBytes;
+
+
+  // Burst length constants
+
+  // Maximal allowed burst length field value, must be positive.
+  parameter int unsigned MaxBurstLenField = 2;
+
+  // Effective max burst length (in number of elements)
+  parameter int unsigned MaxBurstEffLen = 1 << MaxBurstLenField;
```
Comment (vogelpi) [inline rtl/simmem_pkg.sv]: I am not sure I get this. With `MaxBurstLenField = 2` (in AXI, this means a maximum burst length of 3 data beats), you then get an effective number of 4 bits to store the burst length. This seems wrong:
- If your effective burst length is the number of data beats (3), you need 2 bits only.
- If your effective burst length is the number of data beats - 1 (as in the AXI spec), you need 2 bits only.
Please correct me or my comments, please clarify what the effective length is and how it relates to the AXI burst length.

File: rtl/simmem_resp_bank.sv (hunk 3 above)
Comment (vogelpi) [reply rtl/simmem_resp_bank.sv]: This sounds good, but looking at the code, I still see the localparam `XBurstEffLenWidth` being defined and used by many many signals. It is not documented how it differs from `BurstLenWidth` or why it is needed. This is not okay.

File: rtl/simmem_resp_bank.sv (hunk 6 above)
Comment (vogelpi) [reply rtl/simmem_resp_bank.sv]: ?

File: rtl/simmem_resp_bank.sv (hunk 7 above)
Comment (vogelpi) [reply rtl/simmem_resp_bank.sv:519]: ?

File: rtl/simmem_pkg.sv (hunk 13 above)
Comment (flaviens) [reply rtl/simmem_pkg.sv]: Oh no... I have again read the documentation too fast. I had understood MaxBurstEffLen = 1 << MaxBurstLenField, but it's just MaxBurstEffLen = MaxBurstLenField + 1. I'll do the necessary changes. Thank you so much for not letting me do garbage like this :+1:

File: rtl/simmem_resp_bank.sv (hunk 6 above)
Comment (flaviens) [reply rtl/simmem_resp_bank.sv]: Sorry, mixed it with the rsp_cnt and forgat it. Done now!

File: rtl/simmem_pkg.sv (hunk 13 above)
Comment (vogelpi) [reply rtl/simmem_pkg.sv]: Please take your time, read the specs carefully, take notes, make sure you understand it and that it makes sense. This back and forth interaction and changes are really time consuming for you doing the implementation but also for people reviewing your code. Things like this must really be minimized.

**Review by GregAC — APPROVED** [review]
As discussed approved to enable merging for wrapping up of GSoC
